[ContentFeature] de-decouple build specific code from content-features

dbajpeyi opened the pull request:
**Asana Task/Github Issue:** <!-- Link to Asana Task/Github Issue --> https://app.asana.com/0/72649045549333/1209507178584805

## Description

<!--
  Provide a terse summary of what the change is, detailed descriptions should belong in ship reviews or tech designs
-->
1. Move out feature setting parsing logic to a `ConfigFeature` class, which allows outside packages to re-use functionality like `getFeatureSetting`, without having to implement the whole `ContentFeature`
2. Move out all `import.meta` code from the core code, and pass it as `importConfig` in the `content-scope-features.js` class


## Testing Steps


## Checklist

<!--
  These questions are a friendly reminder to shipping code, if you're uncertain ask the AoR owners.
  It's also totally appropriate to not check some of these boxes, if they don't apply to your change.
-->
*Please tick all that apply:*

- [ ] I have tested this change locally
- [ ] I have tested this change locally in all supported browsers
- [ ] This change will be visible to users
- [ ] I have added automated tests that cover this change
- [ ] I have ensured the change is gated by config
- [ ] This change was covered by a ship review
- [x] This change was covered by a tech design
- [ ] Any dependent config has been merged

jonathanKingston commented on injected/entry-points/android.js:
```diff
@@ -31,10 +30,14 @@ function initCode() {
         debug: processedConfig.debug,
     });
 
+    const importConfig = {
```
Maybe move this into content-scope-features to remove the need for it being in all files?

dbajpeyi replied:
Done!

jonathanKingston commented on injected/src/features/web-compat.js:
```diff
@@ -463,7 +463,7 @@ export class WebCompat extends ContentFeature {
     mediaSessionFix() {
         try {
-            if (window.navigator.mediaSession && import.meta.injectName !== 'integration') {
+            if (window.navigator.mediaSession && this.injectName !== 'integration') {
```
Should this be `this.importConfig.injectName`?

dbajpeyi replied:
There's a getter in the base class.

jonathanKingston commented on injected/src/content-feature.js:
```diff
@@ -21,41 +22,196 @@ import { extensionConstructMessagingConfig } from './sendmessage-transport.js';
-        this.#args = null;
+    constructor(featureName, importConfig) {
+        super(featureName);
+        this.args = null;
```
```suggestion
        this.#args = null;
```

jonathanKingston commented on injected/src/content-scope-features.js:
```diff
@@ -39,11 +36,16 @@ export function load(args) {
     for (const featureName of featureNames) {
         const ContentFeature = platformFeatures['ddg_feature_' + featureName];
-        const featureInstance = new ContentFeature(featureName);
+        const featureInstance = new ContentFeature(featureName, importConfig);
         featureInstance.callLoad(args);
```
```suggestion
        const featureInstance = new ContentFeature(featureName, importConfig, args);
        featureInstance.callLoad();
```


Later on:
```
featureInstance.callInit(args);
```

jonathanKingston replied:
This change would allow us to move the initLoadArgs it super constructor.

jonathanKingston commented on injected/src/content-feature.js:
```diff
@@ -21,41 +22,34 @@ import { extensionConstructMessagingConfig } from './sendmessage-transport.js';
-    constructor(featureName) {
-        this.name = featureName;
-        this.#args = null;
+    constructor(featureName, importConfig) {
+        super(featureName);
```
```suggestion
    constructor(featureName, importConfig, args) {
        super(featureName, args);
```

jonathanKingston commented on injected/src/config-feature.js:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * @param {import('./content-scope-features.js').LoadArgs} loadArgs
+     */
+    initLoadArgs(loadArgs) {
+        const { bundledConfig, site, platform } = loadArgs;
+        this.#bundledConfig = bundledConfig;
+        this.#args = loadArgs;
+        // If we have a bundled config, treat it as a regular config
+        // This will be overriden by the remote config if it is available
+        if (this.#bundledConfig && this.#args) {
+            const enabledFeatures = computeEnabledFeatures(bundledConfig, site.domain, platform.version);
+            this.#args.featureSettings = parseFeatureSettings(bundledConfig, enabledFeatures);
+        }
+    }
```
All this moves into the constructor.

dbajpeyi replied:
Moved to config feature constructor!

jonathanKingston commented on injected/src/config-feature.js:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * @param {any} name
+     */
+    constructor(name) {
```
```suggestion
    constructor(name, args) {
```

jonathanKingston commented on injected/src/content-feature.js:
```diff
@@ -144,121 +137,11 @@ export default class ContentFeature {
     callInit(args) {
         const mark = this.monitor.mark(this.name + 'CallInit');
-        this.#args = args;
+        this.args = args;
```
This becomes a method called `setArgs`. That we call from the constructor and `callInit`. We remove these two lines from `callLoad`.

jonathanKingston commented on injected/src/content-feature.js:
```diff
@@ -311,23 +194,15 @@ export default class ContentFeature {
-        }
-        this.#trackerLookup = args.trackerLookup;
-        this.#documentOriginIsTracker = args.documentOriginIsTracker;
+        this.initLoadArgs(args);
```
This initLoadArgs becomes part of the constructor (just calling super with the args).

jonathanKingston commented on injected/src/content-feature.js:
```diff
@@ -311,23 +194,15 @@ export default class ContentFeature {
-        this.#trackerLookup = args.trackerLookup;
-        this.#documentOriginIsTracker = args.documentOriginIsTracker;
+        this.initLoadArgs(args);
         this.load(args);
```
```suggestion
        this.load(this.#args);
```

dbajpeyi commented on injected/src/content-scope-features.js:
```diff
@@ -39,7 +37,7 @@ export function load(args) {
         return;
     }
 
-    const featureNames = typeof import.meta.injectName === 'string' ? platformSupport[import.meta.injectName] : [];
```
Done!

dbajpeyi commented:
@jonathanKingston I've addressed the comments. Note that the references are `this.args` are unique now, because the base class has a setter. Please let me know if it makes sense.

PS: Integration test seems to be broken, but not sure why. I am investigating.

jonathanKingston commented:
@shakyShane this reshuffle starts simplifying some of the application structure. It breaks content-feature into two classes, one which autofill can then depend on or use in isolation.

I suggest when Deep gets back we add more testing on this interface so we can add it as a testable interface rather than just letting this barrier to autofill to rot.

This largely is a transitional step, I don't think we want to reside with args wiring still being so complex (i'd like to break the dependency of features heavily relying on them and instead use methods to access that data).

This also simplifies the load and init arguments mostly as a requirement to remove the use of import.meta which I think we agree now was a bit of a mistake. We just then use it at the periphery and don't require autofill to need custom transpiling. importConfig is the transport object for anything that's import.meta related through into the features.

I've reviewed most of this, I'd appreciate a quick look over in case I missed anything. The 4 additional small tiny commits I made just got the integration tests working from a missed setting of args in a method.

shakyShane commented:
@jonathanKingston I pushed a commit to fix a type, as mentioned in asana I tested this with complex integrations like duck.ai etc.

shakyShane commented:
* **#1521** <a href="https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/1521?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1545** <a href="https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/1545?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/1545?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->
